components/causal_ts/src/observer.rs

```diff
@@ -47,6 +51,25 @@ impl<Ts: CausalTsProvider + 'static> CausalObserver<Ts> {
         coprocessor_host
             .registry
             .register_role_observer(CAUSAL_OBSERVER_PRIORITY, BoxRoleObserver::new(self.clone()));
+        coprocessor_host.registry.register_region_change_observer(
+            CAUSAL_OBSERVER_PRIORITY,
+            BoxRegionChangeObserver::new(self.clone()),
+        );
+    }
+}
+
+const REASON_LEADER_TRANSFER: &str = "leader_transfer";
+const REASON_REGION_MERGE: &str = "region_merge";
+
+impl<Ts: CausalTsProvider> CausalObserver<Ts> {
+    fn flush_timestamp(&self, region: &Region, reason: &'static str) {
+        fail::fail_point!("causal_observer_flush_timestamp", |_| ());
+
+        if let Err(err) = self.causal_ts_provider.flush() {
```

Comment:
So it will be blocked every time a new leader is elected or a region is merged?

Reply:
Yes. It seems to be a performance issue when API v2 is enabled.

The P99 duration of timestamp flush is `1.0 ~ 1.5ms` in a Kingsoft cloud env.
<img width="588" alt="image" src="https://user-images.githubusercontent.com/1907938/171110204-dfc82b42-0cd8-438e-98d8-8ca0c67cf18a.png">

I will think about how to optimize this. As it's not easy to reproduce the causality issues ([#12498](https://github.com/tikv/tikv/issues/12498), [#12680](https://github.com/tikv/tikv/issues/12680)), it may also mean that it is unnecessary to flush timestamp for many circumstances.

How do you think ?

Reply:
> If the ts can be invalidated and flushed asynchronously, then the latency be best optimized.

OK, get it.